src/ERC20Reserve.sol

```diff
@@ -5,54 +5,49 @@ import {Ownable} from "openzeppelin-contracts/access/Ownable.sol";
 import {IERC20} from "openzeppelin-contracts/token/ERC20/IERC20.sol";
 
 interface IERC20Reserve {
-    function erc20() external view returns (IERC20);
+    function withdrawable(IERC20 token) external view returns (uint256 amt);
 
-    function withdraw(uint256 amt) external;
+    function withdraw(IERC20 token, uint256 amt) external;
 
-    function deposit(uint256 amt) external;
+    function deposit(IERC20 token, uint256 amt) external;
 }
 
 contract ERC20Reserve is IERC20Reserve, Ownable {
-    IERC20 public immutable override erc20;
     address public user;
-    uint256 public balance;
+    mapping(IERC20 => uint256) public override withdrawable;
 
     event Withdrawn(address to, uint256 amt);
     event Deposited(address from, uint256 amt);
     event ForceWithdrawn(address to, uint256 amt);
     event UserSet(address oldUser, address newUser);
 
-    constructor(
-        IERC20 _erc20,
-        address owner,
-        address _user
-    ) {
-        erc20 = _erc20;
+    constructor(address owner, address _user) {
         setUser(_user);
         transferOwnership(owner);
     }
 
     modifier onlyUser() {
-        require(_msgSender() == user, "Reserve: caller is not the user");
+        require(msg.sender == user, "Reserve: caller is not the user");
         _;
     }
 
-    function withdraw(uint256 amt) public override onlyUser {
+    function withdraw(IERC20 token, uint256 amt) public override onlyUser {
+        uint256 balance = withdrawable[token];
         require(balance >= amt, "Reserve: withdrawal over balance");
-        balance -= amt;
-        emit Withdrawn(_msgSender(), amt);
-        require(erc20.transfer(_msgSender(), amt), "Reserve: transfer failed");
+        withdrawable[token] = balance - amt;
+        require(token.transfer(msg.sender, amt), "Reserve: transfer failed");
+        emit Withdrawn(msg.sender, amt);
```

Comment:
I think we should add the erc20 `address`  to the deposit, withdraw events

Reply:
Fixed